rfc/2022/2022-09-15_Design-System-Team_Migrate-vets-website-off-of-formation.md

```diff
@@ -0,0 +1,106 @@
+### Prerequisites
+
+1. Application, sitewide, & platform code will need to be using proper VADS web components. This means no React components,
+and no CSS class components as these rely on `formation` components.
+    - This will unblock those modules from being removed in formation
+1. We will need a published version of the `css-library` package
+1. Write a new ESLint rule which can trigger when a `vads-u-*` class is used.
+
+Once these have been met we can move on to beginning the migration
+
+### Migration
+
+In order to begin migrating we will need to make the `css-library` available, and this can be done in one of two ways.
+
+1. Add the stylesheets to the sitewide SASS file.
+    - This will cause every page to download the file whether they use the utilities or not
+    - Could be a "final state" of the migration
+1. Add the stylesheets to each specific app's SASS file as they are migrated.
+    - This will bloat the app's bundle size, but other pages and apps won't be affected
+
+We will be starting with migrating the utility classes, and the css library will export a `utilities.css` file as part of its build which can be imported in either location. All of the steps outlined here will begin with a single app, and then expanded to other apps once the pilot app has been successfully migrated.
+
+#### Remove old VADS utility classes from markup
+We will first identify a team and/or app and work with them on migrating their markup to use the new USWDS3-based utilities provided by `css-library`. As part of this we should also see if we can eliminate any unused CSS from the application's SASS file.
+
+As we eliminate `vads-u-*` classes from app code, add an [ESLint override](https://eslint.org/docs/latest/user-guide/configuring/configuration-files#how-do-overrides-work) to their app directory to throw an error if the older style of utility class is added back into a file.
+
+#### Remove component/module CSS
+
+This is related to the prerequisite for migrating to web components. For example, [this bit of app code](https://github.com/department-of-veterans-affairs/vets-website/blob/4885e8c532f77801712a6d39c0625f8ceb19a556/src/applications/vaos/new-appointment/components/VAFacilityPage/FacilitiesNotShown.jsx#L81) relies on [this CSS from formation](https://github.com/department-of-veterans-affairs/veteran-facing-services-tools/blob/f0e1d666503ecf4aafcb421bbc47fc7f76abec4a/packages/formation/sass/modules/_m-additional-info.scss#L1-L4), and that module is [included in the sitewide styles for `vets-website`](https://github.com/department-of-veterans-affairs/vets-website/blob/4885e8c532f77801712a6d39c0625f8ceb19a556/src/platform/site-wide/sass/style.scss#L16). As we get to a point where we can be sure that certaion SASS modules are no longer being used, we can remove them from `vets-website`, giving us a good idea of how far along we are in the migration while also reducing the bundle size.
+
+#### Try to replace "Custom CSS" with utility classes
```

Comment:
I think we've discussed this before. I'm not so sure this doesn't just move the problem to bloated HTML. If each app only loads their own "custom CSS" (I rankle at the notion of "custom", how about just app-specific CSS?) then at least there is only a penalty of downloading that CSS for a user when they reach that application and applications are isolated from impacting one another. Correct?

Reply:
That's correct _if_ we're just talking about app-specific CSS. There's also site-wide CSS that suffers from this same problem. [Here's an example of that](https://github.com/department-of-veterans-affairs/vets-website/blob/c9bb88731a74b4118842f602fd1a8291ce9d3102/src/platform/site-wide/sass/shame.scss#L37-L39). Every page is getting CSS meant for the dashboard because that CSS exists in the site-wide SASS file.

Even if it does make the markup more bloated, developers have to write less CSS and we don't have the problem where someone adds a CSS class to use in the markup, removes the class from the markup, but leaves the class behind in the CSS file.

Reply:
We should employ a tool in the build process to look for unused CSS and remove it.